plugin/metrics/prometheus/metricsstore/reader_test.go

```diff
@@ -121,16 +142,19 @@ func TestMetricsServerError(t *testing.T) {
 	defer mockPrometheus.Close()
 
 	logger := zap.NewNop()
+	tracer, exp, closer := tracerProvider(t)
 	address := mockPrometheus.Listener.Addr().String()
-	reader, err := NewMetricsReader(logger, config.Configuration{
+	reader, err := NewMetricsReader(config.Configuration{
 		ServerURL:      "http://" + address,
 		ConnectTimeout: defaultTimeout,
-	})
+	}, logger, tracer)
 	require.NoError(t, err)
 
 	m, err := reader.GetCallRates(context.Background(), &params)
+	assert.NotEmpty(t, exp.GetSpans(), "Expected spans are recorded")
+	assert.Len(t, exp.GetSpans(), 1, "HTTP request was traced and span reported")
```

Comment:
Since server is returning an error, I assume the span status should also be error, which we can assert.

Reply:
Do you mean exp.GetSpans()[0].Status ?

Reply:
yes, but not checking it for null, checking it for expected value.
Two other problems:
  1. the asserts are in the wrong place. This test is called TestMetricsServerError. It is testing specific business functionality. Testing spans is a side effect, it should be done later, not as the very first (i.e. most important) thing after calling the business function.
  2. The order of the two asserts doesn't make sense. If the list of spans is empty, then `Spans()[0]` panics, so you never even get to assert.Len(). If you switch the order, it's still incorrect because `assert` does not abort the test, so if the list of spans is empty, then `assert.Len` will log an error and continue, but `Spans()[0]` will panic again. The right way to approach this is to make first check a `require` instead of `assert`. `require` aborts the test if not satisfied.

Reply:
Insightful noted. 🚀